packages/provider/src/MultiChainProvider.test.ts

```diff
@@ -0,0 +1,404 @@
+    typeof ethereum.request
+  >;
+
+  request.mockImplementation(async () => ({
+    jsonrpc: '2.0',
+    id: 1,
+    result: {
+      namespaces: {
+        eip155: getSessionNamespace(),
+      },
+    },
+  }));
+
+  const provider = new MultiChainProvider();
+
+  const { approval } = await provider.connect({
+    requiredNamespaces: namespaces,
+  });
+  await approval();
+
+  return provider;
+}
+
+describe('MultiChainProvider', () => {
+  describe('connect', () => {
+    it('validates the connection arguments', async () => {
+      const provider = new MultiChainProvider();
+
+      // @ts-expect-error Invalid arguments.
+      await expect(provider.connect({ foo: 'bar' })).rejects.toThrow(
+        'Invalid connect arguments',
+      );
+    });
+
+    it('connects', async () => {
+      const request = ethereum.request as jest.MockedFunction<
+        typeof ethereum.request
+      >;
+
+      request.mockImplementation(async () => ({
+        jsonrpc: '2.0',
+        id: 1,
+        result: {
+          namespaces: {
+            eip155: getSessionNamespace(),
+          },
+        },
+      }));
+
+      const provider = new MultiChainProvider();
+      expect(provider.isConnected).toBe(false);
+
+      const { approval } = await provider.connect({ requiredNamespaces: {} });
+      await approval();
+
+      expect(provider.isConnected).toBe(true);
+    });
+
+    it('emits a "session_update" when connected', async () => {
+      const request = ethereum.request as jest.MockedFunction<
+        typeof ethereum.request
+      >;
+
+      request.mockImplementation(async () => ({
+        jsonrpc: '2.0',
+        id: 1,
+        result: {
+          namespaces: {
+            eip155: getSessionNamespace(),
+          },
+        },
+      }));
+
+      const provider = new MultiChainProvider();
+
+      const listener = jest.fn();
+      provider.on('session_update', listener);
+
+      const { approval } = await provider.connect({ requiredNamespaces: {} });
+      await approval();
+
+      expect(listener).toHaveBeenCalledTimes(1);
+      expect(listener).toHaveBeenCalledWith({
+        params: {
+          namespaces: {
+            eip155: getSessionNamespace(),
+          },
+        },
+      });
+    });
+
+    it('defaults to an empty array for the methods and events', async () => {
+      await getProvider({
+        eip155: {
+          chains: ['eip155:1'],
+        },
+      });
+
+      expect(ethereum.request).toHaveBeenCalledWith({
+        method: 'wallet_multiChainRequestHack',
+        params: {
+          jsonrpc: '2.0',
+          id: expect.any(String),
+          method: 'metamask_handshake',
+          params: {
+            requiredNamespaces: {
+              eip155: {
+                chains: ['eip155:1'],
+                methods: [],
+                events: [],
+              },
+            },
+          },
+        },
+      });
+    });
+
+    it('verifies the JSON-RPC response', async () => {
```

Comment:
Is that a user facing requirement?

Reply:
Nope, but regardless a good thing to test I think.